File: Week 2/5_fibonacci_number_again/fibonacci_huge.py
# Uses python3
import sys

def get_pisano_period(n):
    a, b, c = 0, 1, 1
    for _ in range(n*n):
        c = (a + b) % n
        a, b = b, c
        if a == 0 and b == 1:
            return _ + 1


def get_fibonacci_huge(n, m):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    # print("Calculating pisano period...")
    pisano = get_pisano_period(m)
    n = n % pisano
    # print(pisano, n)

    if n < 1:
        return 0
    else:
        for _ in range(n - 1):
            previous, current = current, previous + current

    # print(current)
    return current % m


if __name__ == '__main__':
    n, m = map(int, input().split())
    print(get_fibonacci_huge(n, m))

File: Week 2/5_fibonacci_number_again/fibonacci_huge.cpp
#include <iostream>
// using namespace std;

long long int get_pisano_period(long long int n) {
  int a = 0, b = 1, c = a + b;
  for(long long int i = 0; i < n * n; i++) {
    c = (a + b) % n;
    a = b;
    b = c;
    if(a == 0 & b == 1)
      return i + 1;
  }
  return -1;
}

long long get_fibonacci_huge_naive(long long n, long long m) {
    if (n <= 1)
        return n;

    // std:: cout << "Calculating pisano period...." << "\n";
    long long int pisano = get_pisano_period(m);
    n = n % pisano;
    std::cout << n << " " << pisano << '\n';
    long long previous = 0;
    long long current  = 1;

    for (long long i = 0; i < n - 1; i++) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
    }

    std::cout << current << '\n';
    return current % m;
}

int main() {
    long long n, m;
    std::cin >> n >> m;
    std::cout << get_fibonacci_huge_naive(n, m) << '\n';
}
